### ISSUE TITLE
Issue with inputmask and ckeditor

### ISSUE THREAD

**Issue by RaphaelDDL** [RobinHerbots/Inputmask#319]
Hello there,

I have a form which I use inputmask on some inputs for date/time and a textarea with ckeditor (I don't call inputmask on it).

I discovered that when I call inputmask on an input on IE, Safari and Chrome, ckeditor's jquery adapter (ckeditor on textarea) simply stop working with no warning or anything on console. Opera and Firefox are not affected.

I was reading the jquery adpter code but couldn't find why inputmask messes with it. I made a stackoverflow question on the matter: http://stackoverflow.com/questions/18468393/jquery-inputmask-makes-ckeditor-adapter-to-stop-working

And here is a fiddle to reproduce: http://jsfiddle.net/RaphaelDDL/QjVf8/

Could you help me find if the error is on inputmask overwriting something on the adapter (the $.extend on `val()`, maybe)?

Thanks for any help you can point out.

Regards,
Raphael

**Comment by RaphaelDDL** [comment]
I just had an idea about the `$.extend` and checked `$.fn.val` before and after inserting inputmask. Seems `$.fn.val` is getting overwritten, instead of actually extended (imo, had to hold both extends, no?)

`$.fn.val` before inputmask (chrome):

```
function(d) {
    if (arguments.length) {
        var k = this,
            i = [],
            f = this.each(function() {
                var b = a(this),
                    c = b.data("ckeditorInstance");
                if (b.is("textarea") && c) {
                    var f = new a.Deferred;
                    c.setData(d, function() {
                        f.resolve()
                    });
                    i.push(f.promise());
                    return !0
                }
                return g.call(b, d)
            });
        if (i.length) {
            var b = new a.Deferred;
            a.when.apply(this, i).done(function() {
                b.resolveWith(k)
            });
            return b.promise()
        }
        return f
    }
    var f = a(this).eq(0),
        c = f.data("ckeditorInstance");
    return f.is("textarea") && c ? c.getData() : g.call(f)
}
```

`$.fn.val` after inputmask (chrome):

```
function() {
    var t, i, r;
    return arguments.length == 0 ? (t = n(this), t.data("inputmask") ? t.data("inputmask").autoUnmask ? t.inputmask("unmaskedvalue") : (i = n.inputmask.val.apply(t), i != t.data("inputmask")._buffer.join("") ? i : "") : n.inputmask.val.apply(t)) : (r = arguments, this.each(function() {
        var t = n(this),
            i = n.inputmask.val.apply(t, r);
        return t.data("inputmask") && t.triggerHandler("setvalue.inputmask"), i
    }))
}
```

While on Firefox, `$.fn.val` is always the same (does not change when I call inputmask).

Is there a way I can merge inputmask's extend with ckeditor one?

**Comment by RobinHerbots** [comment]
@RaphaelDDL , 

The n.inputmask.val.apply(t, r); does the call to the original fn.val code.

Probably the ckeditor's adjust the fn.val after the inputmask made the ref in $.inputmask.val.

In FF the patch is not needed.  I will have a look at it and I will also cleanup the val patch to use a valhook instead, which is the way to hook into fn.val provided by jquery.

I will use the jsfiddle to verify.

best regards, 

Robin

**Comment by RaphaelDDL** [comment]
@RobinHerbots 

I call inputmask .js before ckeditor (it's bundled in my minified js, while ckeditor is project specific thus called later). 

> The n.inputmask.val.apply(t, r); does the call to the original fn.val code.
> Probably the ckeditor's adjust the fn.val after the inputmask made the ref in $.inputmask.val.

If I understood correctly what you said, inputmask make the ref when it was inserted onto html (instead of when it's called via `$(selector).inputmask()`). It caches the `$.fn.val` onto `$.inputmask.val` but just actually extend it it when called using `.inputmask()`, right? 
If that, then ckeditor modifies later the val (was called later) but since you got the cached original val from before ckeditor, it overwrites the ckeditor one.

Using this idea as base, I tested moving the ckeditor scripts to before the inputmask script (in fact, before my entire minified js).  I made a rev on the fiddle changing the minified for last http://jsfiddle.net/RaphaelDDL/QjVf8/8/

Seems it worked on the browsers who were showing an overwritten `$.fn.val` now :dancer: 

Your comment gave me sight of what was happening. You are awesome! Thank you very much!

---

Let me know when you do the valHook, so I can test for you, if you require so.
I just read about [what valhook is](http://stackoverflow.com/questions/6731153/what-are-jquery-valhooks) (never heard before). Seems a nice way to extend val/attr without actually modifying the original one, by editing the get/set.
Will totally read more about it.

Best regards,
Raphael

ps.: Sorry for giant text.

**Comment by RobinHerbots** [comment]
@RaphaelDDL, 

I modified the inputmask to use a valhook instead of patching the fn.val.  Can you have a try.

**Comment by RaphaelDDL** [comment]
@RobinHerbots 

http://jsfiddle.net/QjVf8/11/

Using the fiddle I made as base, I removed my minified and added yours.
Seems it worked fine. Will test with more options and stuff to see if broke anything and let you know.

I'm so going to study how you use those valHooks. I might convert the ckeditor adapter to it also.

Thank you very much

Best Regards,
Raphael

P.s.: I just noticed a 'oh my gosh' on every keypress (I've checked the .js history and seems you added/forgot it on #323 [FF fix for backspace](https://github.com/RobinHerbots/jquery.inputmask/blob/019d5679af1d4ee8895c3f1aac50177b965082df/js/jquery.inputmask.js#L1414) :D ).
